packages/aws-cdk-lib/aws-iam/lib/private/imported-role.ts

```diff
@@ -75,7 +75,20 @@ export class ImportedRole extends Resource implements IRole, IComparablePrincipa
   }
 
   public addManagedPolicy(policy: IManagedPolicy): void {
-    Annotations.of(this).addWarningV2('@aws-cdk/aws-iam:importedRoleManagedPolicyNotAdded', `Not adding managed policy: ${policy.managedPolicyArn} to imported role: ${this.roleName}`);
+    // Using "Type Predicate" to confirm x is ManagedPolicy, which allows to avoid
+    // using try ... catch and throw error.
+    const isManagedPolicy = (x: IManagedPolicy): x is ManagedPolicy => {
+      return (x as ManagedPolicy).attachToRole !== undefined;
+    };
+
+    if (isManagedPolicy(policy)) {
+      policy.attachToRole(this);
+    } else {
+      Annotations.of(this).addWarningV2(
+        '@aws-cdk/aws-iam:IRoleCantBeUsedWithIManagedPolicy',
+        'Can\'t combine IRole with IManagedPolicy. Use ManagedPolicy directly.',
```

Comment:
We can add more information like `${policy.managedPolicyArn} to imported role: ${this.roleName}` as in the original code to make it clearer why and where the warning is emitted.

Reply:
Corrected with One Common message.